hooks/terraform_validate.sh

```diff
@@ -43,16 +43,88 @@ function per_dir_hook_unique_part {
   local -a -r args=("$@")
 
   local exit_code
-  local validate_output
+  #
+  # Get hook settings
+  #
+  local retry_once_with_cleanup=false
+
+  IFS=";" read -r -a configs <<< "${HOOK_CONFIG[*]}"
+
+  for c in "${configs[@]}"; do
+
+    IFS="=" read -r -a config <<< "$c"
+    key=${config[0]}
+    value=${config[1]}
+
+    case $key in
+      --retry-once-with-cleanup)
+        retry_once_with_cleanup=$value
+        ;;
+    esac
+  done
 
   common::terraform_init 'terraform validate' "$dir_path" || {
     exit_code=$?
     return $exit_code
   }
 
-  # pass the arguments to hook
-  validate_output=$(terraform validate "${args[@]}" 2>&1)
-  exit_code=$?
+  function do_validate {
+    validate_output=$(terraform validate "${args[@]}" 2>&1)
+    exit_code=$?
+    return $exit_code
+  }
+
+  function parse_validate {
+    # Requires jq
+    local exit_code
+    local validate_output
+    local valid
+    local summary
+
+    validate_output=$(terraform validate -json "${args[@]}" 2>&1)
+    exit_code=$?
+
+    valid="$(jq -rc '.valid' <<< "$validate_output")"
+
+    if [ "$valid" == "true" ]; then
+      return 0
+    fi
+
+    # Pretty-print error information
+    echo "$validate_output" | jq '.diagnostics[]'
+
+    # Parse error message
+    while IFS= read -r error_message; do
+      summary="$(jq -rc '.summary' <<< "$error_message")"
+      case $summary in
+        "missing or corrupted provider plugins")
+          return 10
+          ;;
+        "Module source has changed")
+          return 10
+          ;;
+      esac
+    done < <(jq -rc '.diagnostics[]' <<< "$validate_output")
+    return $exit_code
+  }
+
+  if [ "$retry_once_with_cleanup" = true ]; then
+    parse_validate
+    exit_code=$?
+  else
+    do_validate
+    exit_code=$?
+  fi
+
+  if [ $exit_code -eq 10 ] && [ "$retry_once_with_cleanup" = true ]; then
+    if [ -d .terraform ]; then
+      # Will only be displayed if validation fails again.
+      common::colorify "yellow" "Validation failed. Removing .terraform from: $dir_path"
+      rm -rf .terraform
+      do_validate
```

Comment:
Would it make sense to throw another diagnostic message here about re-validation being done? Sort of this:
```suggestion
      common::colorify "yellow" "Re-validating: $dir_path"
      do_validate
```

Reply:
...just to make it obvious that re-validation is being done and not just the dir has been removed.

Reply:
Agree, thanks for the suggestion